File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.cs
using System;
using System.Collections.Generic;
public class Solution {
    public int solution(int[] topping) {
        int answer = 0;
        int n = topping.Length;
        int[] left = new int[n];
        int[] right = new int[n];
        
        HashSet<int> set = new HashSet<int>();
        
        for(int i = 0; i < n; i++)
        {
            set.Add(topping[i]);
            left[i] = set.Count;
        }
        
        set.Clear();
        
        for(int i = n - 1; i >= 0; i--)
        {
            set.Add(topping[i]);
            right[i] = set.Count;
        }
        
        for(int i = 0; i < n - 1; i++)
        {
            if(left[i] == right[i + 1]) answer++;
        }
        
        return answer;
    }
}

// 자르고 hash에 넣고 카운트 같으면 answer++

File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>

using namespace std;

int solution(vector<int> topping) {
    int answer = 0;
    
    map<int, int> right;
    set<int> left;
    
    for(int t : topping)
    {
        right[t]++;
    }
    
    for(int i = 0; i < topping.size() - 1; i++)
    {
        int t = topping[i];
        
        left.insert(t);
        
        right[t]--;
        if(right[t] == 0)
            right.erase(t);
        
        if(left.size() == right.size())
            answer++;
    }
    return answer;
}